### PR TITLE
Back End Implementation

### PR DESCRIPTION
**Pull request by AnonymousWalker**
Note: the **CoreConverter** class connects the UI to the backend

### CODE REVIEW

File: USFMConverter/Core/CoreConverter.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            var fileFormat = Enum.Parse<FileFormat>(fileFormatName);
+            
+            Project project = BuildProject(viewData);
+            RenderDocument renderer = RenderDocument.GetInstance(fileFormat);
+
+            var usfmDocument = await renderer.LoadUSFMsAsync(project.Files, progressCallback);
+            renderer.Render(project, usfmDocument);
+
+            progressCallback(100); // fills the progress bar
+            await Task.Delay(300); // visible completion before transition
```
Comment (rbnswartz) [inline USFMConverter/Core/CoreConverter.cs]: Adding an artificial delay here in my opinion is bad, I understand that it is to allow the progressbar to catch up, but I don't think delaying half a second to get a progressbar to fully fill up is valuable.

File: USFMConverter/Core/Data/Project.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class Project
+    {
+        public Project()
+        {
+            Files = new List<FileInfo>();
+        }
+
+        public string Name { get; set; } = "";
+        public List<FileInfo> Files { get; }
```
Comment (rbnswartz) [inline USFMConverter/Core/Data/Project.cs]: Any reason why this isn't a string?

File: USFMConverter/Core/Data/Project.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        public FileInfo OutputFile { get; set; }
```
Comment (rbnswartz) [inline USFMConverter/Core/Data/Project.cs:16]: This too? Looks like the only part of this that is being used is the path of the file.

File: USFMConverter/Core/Util/FileSystem.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                    files.Add(file);
+                }
+            }
+
+            return files;
+        }
+
+        /// <exception cref="UnauthorizedAccessException"/>
+        /// <exception cref="DirectoryNotFoundException"/>
+        public static void CheckWritePermission(string path)
```
Comment (rbnswartz) [inline USFMConverter/Core/Util/FileSystem.cs]: In my opinion, if you're going to implement a function like this you should return a boolean to specify whether or not you can write to the file. This adds no value to the application otherwise

File: USFMConverter/App.axaml.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    public class App : Application
+    {
+        public override void Initialize()
+        {
+            AvaloniaXamlLoader.Load(this);
+        }
+
+        public override void OnFrameworkInitializationCompleted()
+        {
+            if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
```
Comment (PurpleGuitar) [inline USFMConverter/App.axaml.cs:17]: Can ApplicationLifetime ever be a different value in this app? It looks like we always start it with Classic in Program.cs.

File: USFMConverter/UI/ViewData.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+namespace USFMConverter.UI
+{
+    public class ViewData
+    {
+        public List<string> Files { get; set; } = new();
+
+        public ComboBoxItem OutputFileFormat { get; set; }
+
+        public int SelectedTextSizeIndex { get; set; } = 0; // binding allows resetting user input
+        public ComboBoxItem TextSize { get; set; }
```
Comment (PurpleGuitar) [inline USFMConverter/UI/ViewData.cs:13]: I recommend against using ComboBoxItems here. This class looks like it's essentially a view model shared between the front and back ends, and the back end shouldn't have to care about the implementation of the front end.  I recommend using Strings instead of ComboBoxItems.

File: USFMConverter/Core/Render/RenderDocument.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+using USFMConverter.Core.ConstantValue;
+using USFMConverter.Core.Data;
+using USFMToolsSharp;
+using USFMToolsSharp.Models.Markers;
+
+namespace USFMConverter.Core.Render
+{
+    public abstract class RenderDocument
+    {
+        public RenderDocument()
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderDocument.cs]: Why empty constructor?

File: USFMConverter/Core/CoreConverter.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            projectBuilder.SetChapterBreak(viewData.ChapterBreak);
+            projectBuilder.SetVerseBreak(viewData.VerseBreak);
+            projectBuilder.EnableNoteTaking(viewData.NoteTaking);
+            projectBuilder.EnableTableOfContents(viewData.TableOfContents);
+            projectBuilder.SetOutputLocation(viewData.OutputFileLocation);
+
+            return projectBuilder.Build();
+        }
+
+        private TextSize GetTextSize(string? sizeName)
```
Comment (rbnswartz) [inline USFMConverter/Core/CoreConverter.cs]: This should have been set a properties with custom getters and setters, your java accent is showing

File: USFMConverter/Core/CoreConverter.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            if (string.IsNullOrEmpty(sizeName))
+            {
+                return TextSize.MEDIUM;
+            }
+
+            return Enum.Parse<TextSize>(sizeName);
+        }
+
+        private LineSpacing GetLineSpacing(string? spacingName)
```
Comment (rbnswartz) [inline USFMConverter/Core/CoreConverter.cs]: This should have been set a properties with custom getters and setters, your java accent is showing

File: USFMConverter/Core/CoreConverter.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            var files = viewData.Files.Select(f => new FileInfo(f));
+
+            var textSizeName = viewData.TextSize.Tag?.ToString();
+            var lineSpacing = viewData.LineSpacing.Tag?.ToString();
+
+            var textAlignment = (viewData.Justified)
+                ? TextAlignment.JUSTIFIED
+                : TextAlignment.LEFT;
+
+            projectBuilder.AddFiles(files.ToList());
```
Comment (rbnswartz) [inline USFMConverter/Core/CoreConverter.cs]: you could have skipped the whole builder here and just set the properties in the constructor. Because you can set properties without specifying them in the constructor at all
```csharp
var project = new Project(){
TextAlign = textAlignment
};

File: USFMConverter/Core/Render/RenderDocx.cs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        };
+
+        public RenderDocx()
+        {
+
+        }
+
+        private DocxConfig BuildDocxConfig(RenderFormat format)
+        {
+            DocxConfig config = new DocxConfig();
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderDocx.cs]: Here you can do this all in the same constructor which I mentioned before

File: USFMConverter/Core/Render/RenderHTML.cs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+using USFMToolsSharp.Models.Markers;
+using USFMToolsSharp.Renderers.HTML;
+
+namespace USFMConverter.Core.Render
+{
+    public class RenderHTML : RenderDocument
+    {
+        private List<string> TextDirectionClasses = new() { "", "rtl-direct" };
+
+        private Dictionary<LineSpacing, string> LineSpacingClasses = new()
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderHTML.cs]: Nice usage of fancy new new()

File: USFMConverter/Core/Render/RenderHTML.cs [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        };
+
+        private Dictionary<TextSize, string> FontSizeClasses = new()
+        {
+            [TextSize.SMALL] = "small-text",
+            [TextSize.MEDIUM] = "med-text",
+            [TextSize.LARGE] = "large-text"
+        };
+
+        public RenderHTML()
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderHTML.cs]: Empty constructor

File: USFMConverter/Core/Render/RenderHTML.cs [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        {
+
+        }
+
+        private HTMLConfig BuildHTMLConfig(RenderFormat format)
+        {
+            HTMLConfig config = new HTMLConfig();
+            var styleClasses = new List<string>();
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderHTML.cs]: you could do something along the lines of 
```csharp
            var styleClasses = new List<string>()
            {
                GetLineSpacingStyle(format.LineSpacing),
                GetColumnStyle(format.ColumnCount),
                GetDirectionStyle(format.LeftToRight),
                GetTextAlignStyle(format.TextAlign),
                GetFontSizeStyle(format.TextSize),
            };
```

File: USFMConverter/Core/Util/FileSystem.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+using System.Threading.Tasks;
+
+namespace USFMConverter.Core.Util
+{
+    public static class FileSystem
+    {
+        public static ICollection<FileInfo> GetFilesInDir(FileInfo dir)
+        {
+            var dirInfo = new DirectoryInfo(dir.FullName);
+            var allFiles = dirInfo.GetFiles("*", SearchOption.AllDirectories);
```
Comment (rbnswartz) [inline USFMConverter/Core/Util/FileSystem.cs]: Directory.GetFiles(dir.FullName, "*", SearchOption.AllDirectories); also works

File: USFMConverter/Core/Render/RenderHTML.cs [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+            <img alt='Creative Commons License' style='border-width:0' src='https://i.creativecommons.org/l/by-sa/4.0/88x31.png' />
+            </p>
+            </div>
+            </td></tr>
+            </table>
+            </div> ";
+            return footerHTML;
+        }
+
+        private string GetLineSpacingStyle(LineSpacing spacing)
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderHTML.cs]: This should have been set a properties with custom getters and setters, your java accent is showing

File: USFMConverter/Core/Render/RenderHTML.cs [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+            return TextDirectionClasses[0];
+        }
+
+        private string GetTextAlignStyle(TextAlignment alignment)
+        {
+            return TextAlignmentClasses[alignment];
+        }
+
+        private string GetFontSizeStyle(TextSize size)
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderHTML.cs]: This should have been set a properties with custom getters and setters, your java accent is showing

File: USFMConverter/Core/Util/FileSystem.cs (hunk 15 above)
Comment (rbnswartz) [reply USFMConverter/Core/Util/FileSystem.cs]: All that you seem to use from FileInfo is getting FullName off of it. More complicated then it needs to be.

File: USFMConverter/Core/Render/RenderDocx.cs [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+            var document = renderer.Render(usfm);
+
+            // check write permission to output location
+            using (Stream outputStream = File.Create(project.OutputFile.FullName))
+            {
+                document.Write(outputStream);
+            }
+        }
+
+        private double GetLineSpacing(LineSpacing spacing)
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderDocx.cs]: This should have been set a properties with custom getters and setters, your java accent is showing

File: USFMConverter/Core/Render/RenderDocx.cs [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            [LineSpacing.SINGLE] = 1.0,
+            [LineSpacing.ONE_HALF] = 1.5,
+            [LineSpacing.DOUBLE] = 2.0,
+            [LineSpacing.TWO_HALF] = 2.5,
+            [LineSpacing.TRIPLE] = 3.0,
+        };
+
+        public RenderDocx()
+        {
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderDocx.cs]: Empty constructor.

File: USFMConverter/Core/Render/RenderDocx.cs [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        public override void Render(Project project, USFMDocument usfm)
+        {
+            var config = BuildDocxConfig(project.FormatOptions);
+            var renderer = new DocxRenderer(config);
+
+            var document = renderer.Render(usfm);
+
+            // check write permission to output location
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderDocx.cs]: This actually doesn't really do a test, this comment is wrong

File: USFMConverter/Core/Render/RenderDocument.cs [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+                
+                // update progress bar
+                var percent = (double)i / totalFiles * 100;
+                progressCallback(percent);
+            }
+
+            return usfmDoc;
+        }
+
+        public void Render(Project project)
```
Comment (rbnswartz) [inline USFMConverter/Core/Render/RenderDocument.cs]: This is dead code and the function it is calling is also dead code

File: USFMConverter/Core/ProjectBuilder.cs [hunk 22]
```diff
@@ -0,0 +1,91 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using USFMConverter.Core.ConstantValue;
+using USFMConverter.Core.Data;
+
+namespace USFMConverter.Core
+{
+    public class ProjectBuilder
```
Comment (rbnswartz) [inline USFMConverter/Core/ProjectBuilder.cs]: This whole entire thing is a good idea but way too overcomplicated, if you have a class that have functions that are one line long just setting a property on another object, it is way too overcomplicated

File: USFMConverter/Core/Util/FileSystem.cs [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace USFMConverter.Core.Util
+{
+    public static class FileSystem
+    {
+        public static ICollection<FileInfo> GetFilesInDir(FileInfo dir)
```
Comment (rbnswartz) [inline USFMConverter/Core/Util/FileSystem.cs]: Why does this function exist? It is just a very small wrapper around GetFiles that doesn't add any value it just makes things more complicated.

File: USFMConverter/Core/CoreConverter.cs (hunk 1 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/CoreConverter.cs]: I think this should be a Thomas question about UX. In my opinion, it's nicer to visually catch up with a full progress bar before showing the success dialog and that 300ms doesn't cost much as far as I know

File: USFMConverter/Core/CoreConverter.cs (hunk 10 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/CoreConverter.cs]: I think it's better to use methods of builder pattern since we don't want to have 10+ constructor parameters which isn't a good practice

File: USFMConverter/Core/ProjectBuilder.cs (hunk 22 above)
Comment (PurpleGuitar) [inline USFMConverter/Core/ProjectBuilder.cs]: This class seems superfluous.  It doesn't seem to add any value over a constructor for the Project class.  I recommend removing it.

File: USFMConverter/Core/Data/Project.cs (hunk 3 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/Data/Project.cs:16]: I have one case that uses the parent directory in Core/Render/RenderHTML.cs

```
        public override void Render(Project project, USFMDocument usfmDoc)
        {
            ...
            var cssFilename = Path.Combine(project.OutputFile.DirectoryName!, "style.css");
            ...
        }
```

File: USFMConverter/Core/Render/RenderDocument.cs [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using USFMConverter.Core.ConstantValue;
+using USFMConverter.Core.Data;
+using USFMToolsSharp;
+using USFMToolsSharp.Models.Markers;
+
+namespace USFMConverter.Core.Render
+{
+    public abstract class RenderDocument
```
Comment (PurpleGuitar) [inline USFMConverter/Core/Render/RenderDocument.cs]: This class name is confusing.  A class name should be a noun that states what it is.  "RenderDocument" is a verb, not a noun, and makes it unclear what this is.  Instead, this should be named something like "RenderableDocument".

File: USFMConverter/Core/Render/RenderDocx.cs [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+using System.Collections.Generic;
+using System.IO;
+using USFMConverter.Core.ConstantValue;
+using USFMConverter.Core.Data;
+using USFMToolsSharp.Models.Markers;
+using USFMToolsSharp.Renderers.Docx;
+
+namespace USFMConverter.Core.Render
+{
+    public class RenderDocx : RenderDocument
```
Comment (PurpleGuitar) [inline USFMConverter/Core/Render/RenderDocx.cs]: This class is confusingly named.  It should be named something like "DocxDocument".

File: USFMConverter/Core/Render/RenderHTML.cs [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+using System.IO;
+using System.Linq;
+using USFMConverter.Core.ConstantValue;
+using USFMConverter.Core.Data;
+using USFMToolsSharp.Models.Markers;
+using USFMToolsSharp.Renderers.HTML;
+
+namespace USFMConverter.Core.Render
+{
+    public class RenderHTML : RenderDocument
```
Comment (PurpleGuitar) [inline USFMConverter/Core/Render/RenderHTML.cs]: This class is confusingly named.  It should be called something like "HTMLDocument."

File: USFMConverter/Core/ProjectBuilder.cs (hunk 22 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/ProjectBuilder.cs]: Yeah I agree, we could set it directly in the convert method

File: USFMConverter/Core/Render/RenderDocument.cs [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            
+        }
+
+        /// <summary>
+        /// Parses the given text files into one USFM Document.
+        /// </summary>
+        /// <param name="files">Text files with USFM format.</param>
+        /// <returns>A USFM Document</returns>
+        public USFMDocument LoadUSFMs(IEnumerable<FileInfo> files)
```
Comment (PurpleGuitar) [inline USFMConverter/Core/Render/RenderDocument.cs]: These functions don't belong here -- they have nothing specifically to do with a RenderableDocument.  They should either be in a utility class such as USFMUtils, or perhaps in the FileSystem interface.

File: USFMConverter/Core/Render/RenderDocument.cs (hunk 21 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/Render/RenderDocument.cs]: I think it's nice to have both sync and asyn loading, not wasting so much resource though

File: USFMConverter/Core/Render/RenderDocument.cs [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+            return usfmDoc;
+        }
+
+        /// <summary>
+        /// Parses the given text files into one USFM Document asynchronously.
+        /// </summary>
+        /// <param name="files">Text files with USFM format.</param>
+        /// <param name="progressCallback">Call back for progress bar update.</param>
+        /// <returns>A USFM Document</returns>
+        public async Task<USFMDocument> LoadUSFMsAsync(
```
Comment (PurpleGuitar) [inline USFMConverter/Core/Render/RenderDocument.cs]: This function is not related to the RenderableDocument class.  It should be move into a utilities or file interface class.

File: USFMConverter/Core/Render/RenderDocument.cs (hunk 7 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/Render/RenderDocument.cs]: removed!

File: USFMConverter/Core/Data/Project.cs (hunk 2 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/Data/Project.cs]: I am not sure I have a good reason to use string.

File: USFMConverter/Core/Render/RenderDocument.cs [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        public void Render(Project project)
+        {
+            Render(project, LoadUSFMs(project.Files));
+        }
+
+        public abstract void Render(Project project, USFMDocument usfmDoc);
+
+        public static RenderDocument GetInstance(FileFormat fileFormat)
```
Comment (PurpleGuitar) [inline USFMConverter/Core/Render/RenderDocument.cs]: A superclass should not be making direct reference to subclasses.  This is an appropriate place to use a Builder class.

File: USFMConverter/Core/Render/RenderDocx.cs (hunk 11 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/Render/RenderDocx.cs]: I think it's better to have this method acting as a function, so that when the parameter changes, we don't need to construct another render object

**Review by PurpleGuitar — CHANGES_REQUESTED** [review]

File: USFMConverter/Core/Util/FileSystem.cs (hunk 23 above)
Comment (AnonymousWalker) [reply USFMConverter/Core/Util/FileSystem.cs]: since it may not be appropriate for the UI code having back end storage access

File: USFMConverter/App.axaml.cs (hunk 5 above)
Comment (AnonymousWalker) [reply USFMConverter/App.axaml.cs:17]: I believe it's the Avalonia generated code

File: USFMConverter/UI/ViewData.cs (hunk 6 above)
Comment (AnonymousWalker) [reply USFMConverter/UI/ViewData.cs:13]: I agree with your idea but we had one problem with that. Since we use view data binding to reduces the unnecessary event handler for every input controls, there is a limitation to that. The ComboBox (dropdown) seems to have 2 useful property that binding can make sense: SelectedIndex and SelectedItem. We chose the SelectedItem to bind because index is not reliable. Thus, the value of SelectedItem is whatever object it received from the drop drown selection - ComboBoxItem in this case. We embed a value to its "Tag" property so that we can collect at the end.

And by limiting the use of this ViewData to the front end and CoreConverter.cs, it will be mapped/adapted to the appropriate values within the converter only, thus it won't leak into the back end code.

File: USFMConverter/App.axaml.cs (hunk 5 above)
Comment (PurpleGuitar) [reply USFMConverter/App.axaml.cs:17]: OK, if it's boilerplate I guess it's fine.

File: USFMConverter/UI/ViewData.cs (hunk 6 above)
Comment (PurpleGuitar) [reply USFMConverter/UI/ViewData.cs:13]: I suppose it's a price worth paying for the automatic binding.

Comment (AnonymousWalker) [reply USFMConverter/UI/ViewData.cs:13]: @rbnswartz feel free to let us know your thought, too

**Review by PurpleGuitar — APPROVED** [review]

**Comment by AnonymousWalker** [comment]
Updated changes are committed to the child PR https://github.com/Bible-Translation-Tools/USFM-Converter/pull/30

File: USFMConverter/Core/CoreConverter.cs (hunk 10 above)
Comment (rbnswartz) [reply USFMConverter/Core/CoreConverter.cs]: You don't need to add them as a parameter, you can specify them after the constructor like I'm doing above. The constructor doesn't have to do anything.

Comment (AnonymousWalker) [reply USFMConverter/Core/CoreConverter.cs]: okay I made the change in the child PR (#30)

File: USFMConverter/Core/Data/Project.cs (hunk 2 above)
Comment (rbnswartz) [reply USFMConverter/Core/Data/Project.cs]: Everywhere you are using it you just cast it to a string anyway. Unless I missed something where you are using more then just the full path name. Also every time you need to add something into this (which you are doing from a string) it needs to be converted to a fileinfo.
